ads/ads.py

```diff
@@ -312,6 +321,80 @@ def __init__(self, name, selector_set):
         self.selectors = selector_set
 
 
+##############################################
+# Cache
+##############################################
+
+ADS_ROOT = "adsroot"
+
+class Cache:
+    @classmethod
+    def get_cache_path(cls, dir_):
+        adscache = ".ads_cache.yml"
+        cache_home = os.getenv("ADS_CACHE_HOME")
+        if cache_home:
+            return "%s/%s" % (cache_home, adscache)
+
+        return "%s/%s" % (dir_, adscache)
+
+    @classmethod
+    def load_from_cache(cls, cachefile, project_file, profile_dir):
+        map_ = {}
+        if os.path.isfile(cachefile):
+            cache_spec = _load_spec_file(cachefile)
+
+            if cache_spec.get(ADS_ROOT) == project_file:
+                del cache_spec[ADS_ROOT]
+                map_ = cache_spec
+
+        return map_
+
+    def __init__(self, project_file, profile_dir):
+        self.cachefile = Cache.get_cache_path(profile_dir)
+        self.cache_map = Cache.load_from_cache(self.cachefile, project_file, profile_dir)
+
+    def get(self, key):
+        if isinstance(key, Service):
+            return self.cache_map.get(str(key))
+
+        return self.cache_map.get(key)
+
+    def valid_groups(self, service_sets):
+        selectors = set()
+        for service_set in service_sets:
+            for selector in service_set.selectors:
+                selectors.add(selector)
+
+        groups = set(group.name for group in service_sets)
+        services = selectors.difference(groups)
+
+        return self.yamls_exist(list(services))
+
+    def yamls_exist(self, services):
+        not_cached = []
+        for service in services:
+            yaml = self.get(service)
+
+            if not yaml or not os.path.isfile(yaml):
+                not_cached.append(service)
+
+        if not_cached:
+            warning("The following service(s) are not cached: "
```

Comment:
I don't get it. What did the user do wrong, and what do they need to do to fix it?